dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,352 @@
+
+
+class DnsPolicyCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: typing.Any):
+        """Construct.
+
+        Args:
+            args: Arguments passed to the CharmBase parent constructor.
+        """
+        super().__init__(*args)
+
+        self.on.define_event("reconciliate", ReconciliateEvent)
+        logger.debug("unit name: %s", self.unit.name)
+
+        self.dns_policy = DnsPolicyService()
+        self._database = DatabaseHandler(self, constants.DATABASE_RELATION_NAME)
+        self.dns_record_provider = dns_record.DNSRecordProvides(self, "dns-record-provider")
+        self.dns_record_requirer = dns_record.DNSRecordRequires(self, "dns-record-requirer")
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(
+            self._database.database.on.database_created, self._on_database_created
+        )
+        self.framework.observe(
+            self._database.database.on.endpoints_changed, self._on_database_endpoints_changed
+        )
+        self.framework.observe(
+            self.on[constants.DATABASE_RELATION_NAME].relation_broken,
+            self._on_database_relation_broken,
+        )
+        self.framework.observe(self.on.create_reviewer_action, self._on_create_reviewer_action)
+        self.framework.observe(
+            self.dns_record_provider.on.dns_record_request_received,
+            self._on_dns_record_request_received,
+        )
+        self.framework.observe(self.on.reconciliate, self._on_reconciliate)
+        self.unit.open_port("tcp", 8080)  # dns-policy-app
+
+    def _on_reconciliate(self, _: ReconciliateEvent) -> None:
+        relation_data = self._get_remote_relation_data()
+        if relation_data is None:
+            return
+        dns_policy_config = self.dns_policy.get_config()
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(entry)
+        # logger.debug("Entries: %s", entries)
+        if not entries:
+            return
+        logger.debug("Config: %s", dns_policy_config)
+        if not self.model.unit.is_leader():
+            return
+
+        token = self.dns_policy.get_api_root_token()
+        self.dns_policy.send_requests(token, entries)
+
+        approved_requests = self.dns_policy.get_approved_requests(token)
+        dns_record_requirer_data = dns_record.DNSRecordRequirerData(dns_entries=approved_requests)
+        for relation in self.model.relations[self.dns_record_requirer.relation_name]:
+            logger.debug("dnsrecordrequirerdata: %s", dns_record_requirer_data)
+            self.dns_record_requirer.update_relation_data(relation, dns_record_requirer_data)
```

Comment:
Won't it put record information of all the relations to every relation databag ?
Shouldn't we filter and only put the data corresponding to the relation

Reply:
There is only one possible `dns_record_requirer` relation (the subordinate one with `bind`)

Reply:
I searched around and used the `get_relation()` function with guards to handle the fact that the relation should exist and be unique.